### PR TITLE
Add SemanticSegmentationEvaluator

### PR DESCRIPTION
**Pull request by yuyu2172**
merge after #217.

The current evaluate.py is wrong because n_positive adds up for pixels with "ignore" GT labels.
I made the following modification.
```
    for batch in it:
        img, gt_label = concat_examples(batch, args.gpu)
        img = chainer.Variable(img, volatile=True)
        pred_label = F.argmax(F.softmax(model(img)), axis=1)
        pred_label = cuda.to_cpu(pred_label.data)
        gt_label = cuda.to_cpu(gt_label)
        for cls_i in range(n_class):
            if cls_i in ignore_labels:
                continue
            pred_label = pred_label[gt_label != -1]
            gt_label = gt_label[gt_label != -1]
            n_positive[cls_i] += np.sum(pred_label == cls_i)
            n_true[cls_i] += np.sum(gt_label == cls_i)
            n_true_positive[cls_i] += np.sum(
                (pred_label == cls_i) * (gt_label == cls_i))
```

I set batchsize to one. This was critical to reproduce the results with SemanticSegmentationEvaluator.

```
                    Sky : 0.878988525
               Building : 0.668390287
                   Pole : 0.192338392
                   Road : 0.873874189
               Pavement : 0.642089768
                   Tree : 0.622743688
             SignSymbol : 0.189341867
                  Fence : 0.213735296
                    Car : 0.635499263
             Pedestrian : 0.273885917
              Bicyclist : 0.241545763
==================================
               mean IoU : 0.493857541
 Class average accuracy : 0.670468304
Global average accuracy : 0.826623129
```

With SemanticSegmentationEvaluator, the scores were as follows. This is identical to the score above.
```
                    Sky : 0.878988525
               Building : 0.668390287
                   Pole : 0.192338392
                   Road : 0.873874189
               Pavement : 0.642089768
                   Tree : 0.622743688
             SignSymbol : 0.189341867
                  Fence : 0.213735296
                    Car : 0.635499263
             Pedestrian : 0.273885917
              Bicyclist : 0.241545763
==================================
               mean IoU : 0.493857541
 Class average accuracy : 0.670468304
Global average accuracy : 0.826623129
```

### CODE REVIEW

**Comment by yuyu2172** [comment]
Since the content of this PR is not purely about `SemanticSegmentationEvaluator`, I split irrelevant part into #249.

**Comment by yuyu2172** [comment]
I changed directory structure of `extensions`. This is the reason why I added `no-compat` label.

**Comment by yuyu2172** [comment]
@Hakuyume 
Can you check this?

**Comment by Hakuyume** [comment]
@yuyu2172 OK

File: chainercv/extensions/evaluator/semantic_segmentation_evaluator.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+from chainercv.evaluations import eval_semantic_segmentation
+from chainercv.utils import apply_prediction_to_iterator
+
+
+class SemanticSegmentationEvaluator(chainer.training.extensions.Evaluator):
+
+    """An extension that evaluates a semantic segmentation model.
+
+    This extension iterates over an iterator and evaluates the prediction
+    results of the model by common evaluation metrics for semantic
```
Comment (Hakuyume) [inline chainercv/extensions/evaluator/semantic_segmentation_evaluator.py]: 'a/the common evaluation metrics`? (I'm not sure which is better)

File: chainercv/extensions/evaluator/semantic_segmentation_evaluator.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+    .. seealso::
+        :func:`chainercv.evaluations.eval_semantic_segmentation`.
+
+    Args:
+        iterator (chainer.Iterator): An iterator. Each sample should be
+            following tuple :obj:`img, label`.
+            :obj:`img` is an image, :obj:`label` is pixel-wise label.
+        target (chainer.Link): A semantic segmentation link. This link should
+            have :meth:`predict` method which takes a list of images and
```
Comment (Hakuyume) [inline chainercv/extensions/evaluator/semantic_segmentation_evaluator.py]: `which` -> `that`? https://github.com/chainer/chainercv/pull/229#discussion_r119510618

File: chainercv/extensions/evaluator/semantic_segmentation_evaluator.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            returns :obj:`labels`.
+        label_names (iterable of strings): An iterable of names of classes.
+            If this value is specified, IoU and class accuracy for each class
+            is also reported with the keys
```
Comment (Hakuyume) [inline chainercv/extensions/evaluator/semantic_segmentation_evaluator.py]: `is` -> `are`

File: docs/source/reference/extensions.rst [hunk 4]
```diff
@@ -4,13 +4,21 @@ Extensions
...
+Evaluator
 ---------
 
-DetectionVisReport
-~~~~~~~~~~~~~~~~~~
-.. autofunction:: DetectionVisReport
-
 DetectionVOCEvaluator
 ~~~~~~~~~~~~~~~~~~~~~
 .. autofunction:: DetectionVOCEvaluator
```
Comment (Hakuyume) [inline docs/source/reference/extensions.rst]: I think it is a class (not a function)

File: tests/extensions_tests/evaluator_tests/test_semantic_segmentation_evaluator.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+        # No observation is reported to the current reporter. Instead the
+        # evaluator collect results in order to calculate their mean.
+        np.testing.assert_equal(len(reporter.observation), 0)
+
+        np.testing.assert_equal(eval_['target/miou'], 1.)
+        np.testing.assert_equal(eval_['target/pixel_accuracy'], 1.)
+        np.testing.assert_equal(eval_['target/mean_class_accuracy'], 1.)
+        np.testing.assert_equal(eval_['target/iou/a'], 1.)
+        np.testing.assert_equal(eval_['target/iou/b'], 1.)
```
Comment (Hakuyume) [inline tests/extensions_tests/evaluator_tests/test_semantic_segmentation_evaluator.py]: Could you make `iou/a` and `iou/b` different from each other?

File: docs/source/reference/extensions.rst (hunk 4 above)
Comment (yuyu2172) [inline docs/source/reference/extensions.rst]: We've been using `autofunction` for many classes (e.g. Dataset class).

Comment (yuyu2172) [inline docs/source/reference/extensions.rst]: They should be fixed too.

Comment (yuyu2172) [inline docs/source/reference/extensions.rst]: #308

**Comment by yuyu2172** [comment]
Thanks for reviewing.
